src/java.base/share/classes/java/nio/file/SecureDirectoryStream.java

```diff
@@ -135,6 +135,182 @@ SeekableByteChannel newByteChannel(T path,
+     * for {@linkplain Files#createDirectory Files.createDirectory}.
+     *
+     * @param   dir
+     *          the path of the directory to create
+     * @param   attrs
+     *          an optional list of file attributes to set atomically when
+     *          creating the directory
+     *
+     * @return  the directory
+     *
+     * @throws  ClosedDirectoryStreamException
+     *          if the directory stream is closed
+     * @throws  UnsupportedOperationException
+     *          if the array contains an attribute that cannot be set atomically
+     *          when creating the directory
+     * @throws  FileAlreadyExistsException
+     *          if a directory could not otherwise be created because a file of
+     *          that name already exists <i>(optional specific exception)</i>
+     * @throws  IOException
+     *          if an I/O error occurs or the parent directory does not exist
+     *
+     * @since 26
+     */
+    Path createDirectory(Path dir, FileAttribute<?>... attrs)
+        throws IOException;
+
+    /**
+     * Creates a new link (directory entry) for an existing file <i>(optional
+     * operation)</i>.
+     *
+     * <p>This method works in a similar manner to {@linkplain Files#createLink
+     * Files.createLink}.  If the {@code link} parameter is an {@link
+     * Path#isAbsolute absolute} path then it locates the link file. If the
+     * {@code link} parameter is a relative path then it is located relative to
+     * this open directory. If the {@code existing} parameter is an absolute
+     * path then it locates the target file (the {@code targetdir} parameter is
+     * ignored). If the {@code existing} parameter is a relative path it is
+     * located relative to the open directory identified by the {@code
+     * targetdir} parameter.
+     *
+     * @param   link
+     *          the link (directory entry) to create
+     * @param   targetdir
+     *          the destination directory
+     * @param   existing
+     *          a path to an existing file
+     *
+     * @return  the path to the link (directory entry)
+     *
+     * @throws  ClosedDirectoryStreamException
+     *          if the directory stream is closed
+     * @throws  UnsupportedOperationException
+     *          if the implementation does not support adding an existing file
+     *          to a directory
+     * @throws  FileAlreadyExistsException
+     *          if the entry could not otherwise be created because a file of
+     *          that name already exists <i>(optional specific exception)</i>
+     * @throws  NoSuchFileException
+     *          if the file specified by the combination of {@code targetdir}
+     *          and {@code existing} does not exist
+     * @throws  IOException
+     *          if an I/O error occurs
+     *
+     * @since 26
+     */
+    Path createLink(T link, SecureDirectoryStream<T> targetdir, T existing)
+        throws IOException;
```

Comment:
I realize this models a pair of dfd/path tuples and mklinkat but I wonder about the usability and whether we really need this. In other, maybe we should drop this method from the proposal for now.

Reply:
Could do, but left in for the moment.

Reply:
For what it's worth, as the author of JDK-8356493, I would use `createLink`.

Maybe the usability of simple cases would be improved by a `createLink(T link, T target)` overload (that passed `olddirfd=AT_FDCWD` to the overlying syscall)?

I think this may also need a `LinkOption` flag to pass `AT_FOLLOW_SYMLINKS`?

Reply:
> Maybe the usability of simple cases would be improved by a `createLink(T link, T target)` overload (that passed `olddirfd=AT_FDCWD` to the underlying syscall)?

This could also be accomplished by passing as `targetdir` a `SecureDirectoryStream` created for the `Path` of `System.getProperty("user.dir")`.

I suppose it depends on how common the case is as to whether a convenience method is worthwhile.

Reply:
I agree that including `AT_SYMLINK_FOLLOW` by default and optionally taking `LinkOption.NOFOLLOW_LINKS` would be most consistent with the rest of the API.